src/scroll-timeline-css.js

```diff
@@ -0,0 +1,192 @@
+  if (matchesSelector) {
+    if (document.getElementById(matchesSelector[1])) {
+      return {
+        target: document.getElementById(matchesSelector[1]),
+        ...(matchesSelector.length >= 3 ? { edge: matchesSelector[2] } : {}),
+        ...(matchesSelector.length >= 4 ? { threshold: parseFloat(matchesSelector[3]) } : {})
+      };
+    }
+  }
+
+  return null;
+}
+
+function isDescendant(child, parent) {
+  while (child) {
+    if (child == parent) return true;
+    child = child.parentNode;
+  }
+  return false;
+}
+
+function getScrollOffsets(source, offsets) {
+  let scrollOffsets = undefined;
+
+  if (offsets) {
+    if (offsets == "none") {
+      // do nothing
+    } else {
+      scrollOffsets = offsets.split(",")
+        .map(part => part.trim())
+        .filter(part => part != '')
+        .map(part => convertOneScrollOffset(part))
+        .filter(offset => offset);
+
+      // If the available scroll-offsets are not valid,
+      // so we won't create the scrollTimeline,
+      for (let off of scrollOffsets) {
+        if (off.target && off.target instanceof Element &&
+          (window.getComputedStyle(off.target, null).display == "none" || !isDescendant(off.target, source))) {
+          return null;
+        }
+      }
+      if (scrollOffsets.length == 0) {
+        return null;
+      }
+    }
+  }
+
+  return scrollOffsets;
+}
+
+function createScrollTimeline(name) {
+  const options = parser.scrollTimelineOptions.get(name);
+  if (!options) return null;
+
+  const sourceElement = getSourceElement(options.source);
+  const scrollOffsets = getScrollOffsets(sourceElement, options['scroll-offsets']);
+
+  if (scrollOffsets !== null) {
+    const scrollTimeline = new ScrollTimeline({
+      ...(sourceElement ? { source: getSourceElement(options.source) } : {}),
+      ...(scrollOffsets ? { scrollOffsets: scrollOffsets } : {}),
+      ...(options.orientation != "auto" ? { orientation: options.orientation } : {}),
+    });
+    return scrollTimeline;
+  } else {
+    return null;
+  }
+}
+
+export function initCSSPolyfill() {
+  initMutationObserver();
+
+  window.addEventListener('animationstart', (evt) => {
+    const anim =
+      evt.target.getAnimations().filter(
+        anim => anim.animationName == evt.animationName)[0];
+
+    const timelineName = parser.getScrollTimelineName(anim.animationName, evt.target);
+
+    if (timelineName) {
+      const scrollTimeline = createScrollTimeline(timelineName);
+      // If there is a scrollTimeline name associated to this animation,
+      // cancel it, whether we create a new animation or not
+      // depends on the fact that whether that scrollTimeline was valid or not
+      anim.cancel();
+      if (scrollTimeline) {
+        if (anim.timeline != scrollTimeline) {
+          const target = anim.effect.target;
+          const keyframes = anim.effect.getKeyframes();
+          target.animate(keyframes, { timeline: scrollTimeline });
```

Comment:
new ProxyAnimation(anim, scrollTimeline);

This way the native animation is still a CSS animation. Furthermore, we are not potentially creating a cancel event or a new animation start event.